### PR TITLE
[ZEPPELIN-2817] Support default interpreter setting in create note re…

### PR DESCRIPTION
**Pull request by reminia**
…st api

### What is this PR for?
Can set default interpreter in create note rest api.

### What type of PR is it?
[Improvement]

### Todos
None

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-2817

### How should this be tested?
Post name and interpreter to rest api/notebook, a note with the interpreter should be created.

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes

### CODE REVIEW

**Comment by reminia** [comment]
Add doc and some modify. @felixcheung @sravan-s , plz help review again.
Some details:

1. post interpreter name not id, cause name is user friendly and id is something like CRB87NS
2. if the interpreter name doesnt exist,  return **bad request status** and no note is created.
3. if successfully created the note, return CREATED status  like the old DOC says.
4. name/interpreter/paragraphs no need to be supplied together, empty message is also ok.

**Comment by reminia** [comment]
@felixcheung @Leemoonsoo  can you help review this?

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java [hunk 1]
```diff
@@ -345,16 +370,20 @@ public Response createNote(String message) throws IOException {
...
+        noteName = request.getName();
+      }
     }
 
     note.setName(noteName);
     note.persist(subject);
     notebookServer.broadcastNote(note);
     notebookServer.broadcastNoteList(subject, SecurityUtils.getRoles());
-    return new JsonResponse<>(Status.OK, "", note.getId()).build();
+    return new JsonResponse<>(Status.CREATED, "", note.getId()).build();
```
Comment (felixcheung) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: if I read this right, the doc says it should be 200 OK
https://github.com/apache/zeppelin/pull/2508/files#diff-ccd7e2a2e181424ea2f4e346189e7688R93
?

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java [hunk 2]
```diff
@@ -334,7 +335,31 @@ public Response createNote(String message) throws IOException {
...
+          break;
+        }
+      }
+      if (defaultId != null) {
+        interpreterIDs.remove(defaultId);
+        interpreterIDs.add(0, defaultId);
+        note = notebook.createNote(interpreterIDs, subject);
+      } else {
+        String errorMsg = String.format("interpreter %s doesn't exist", interpreter);
+        return new JsonResponse<>(Status.BAD_REQUEST, errorMsg).build();
```
Comment (felixcheung) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: we should document this response code
but this should be 404 Not Found instead?

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java (hunk 1 above)
Comment (reminia) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: @felixcheung , I mod the success code in doc to 201.  Cause it's a created api. And also the response json in old doc is CREATED status. [see](https://zeppelin.apache.org/docs/0.8.0-SNAPSHOT/usage/rest_api/notebook.html#create-a-new-note).

**Comment by reminia** [comment]
Test added @felixcheung

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java (hunk 1 above)
Comment (felixcheung) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: I think we need to be super careful with changing an existing response code for an existing API - there is likely existing usage that could be broken with a simple code change.

I get that 201 might be better and/or more consistent with other APIs though. Do you think it makes a difference if we change it now, or we leave it as before?

File: zeppelin-server/src/test/java/org/apache/zeppelin/rest/NotebookRestApiTest.java [hunk 3]
```diff
@@ -207,4 +210,51 @@ public void testClearAllParagraphOutput() throws IOException {
...
     ZeppelinServer.notebook.removeNote(note.getId(), anonymous);
   }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void testCreateNote() throws IOException {
+    // prepare data
+    Map<String, Object> data = Maps.newHashMap();
+    data.put("name", "test");
+    data.put("interpreter", "spark");
```
Comment (felixcheung) [inline zeppelin-server/src/test/java/org/apache/zeppelin/rest/NotebookRestApiTest.java]: perhaps we should test with a different interpreter? `spark` is likely the default even when this isn't set in the request

Comment (reminia) [inline zeppelin-server/src/test/java/org/apache/zeppelin/rest/NotebookRestApiTest.java]: change spark to python interpreter.

**Comment by reminia** [comment]
I decide to leave the api no change at now.  Maybe another jira should be created to address this. I've found that several create operation api doesnt return 201 but 200， such as repo add, paragraph add. @felixcheung @Leemoonsoo

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java [hunk 4]
```diff
@@ -334,7 +335,31 @@ public Response createNote(String message) throws IOException {
...
+      String defaultId = null;
+      for (String id : interpreterIDs) {
+        if (manager.get(id).getName().equals(interpreter)) {
+          defaultId = id;
+          break;
+        }
+      }
+      if (defaultId != null) {
+        interpreterIDs.remove(defaultId);
+        interpreterIDs.add(0, defaultId);
```
Comment (zjffdu) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: Would this cause 2 duplicated `interpreterId` in this `interpreterIDs` ?

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java [hunk 5]
```diff
@@ -345,9 +370,13 @@ public Response createNote(String message) throws IOException {
       }
     }
     note.addNewParagraph(subject); // add one paragraph to the last
-    String noteName = request.getName();
-    if (noteName.isEmpty()) {
-      noteName = "Note " + note.getId();
+
+    //deal with possible empty note name
+    String noteName = "Note " + note.getId();
```
Comment (zjffdu) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: Another option is to do in frontend that we can enforce user to provide a note name ? Personally, I prefer to do it in frontend to enforce user providing note name.

File: docs/usage/rest_api/notebook.md [hunk 6]
```diff
@@ -92,18 +93,31 @@ Notebooks REST API supports the following operations: List, Create, Get, Delete,
...
+  "name": "name of new note",
+  "interpreter": "python"
+}</pre></td>
+    </tr>
+    <tr>
+      <td> sample JSON input (with interpreter, initial paragraphs) </td>
       <td><pre>
 {
   "name": "name of new note",
+  "interpreter": "python",
```
Comment (zjffdu) [inline docs/usage/rest_api/notebook.md]: defaultInterpreter is better for me. `interpreter` is a little misleading.

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java (hunk 5 above)
Comment (reminia) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: Hey Jeff, this is rest api without frontend check. Now frontend can create note without name supplied. This api just tries to stay the same with the web function.

File: zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java (hunk 4 above)
Comment (reminia) [inline zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java]: sorry, I didnt get your idea. Do you mean interpreterID duplicate test should be added? Why?Remove then add , no duplicate.

File: docs/usage/rest_api/notebook.md (hunk 6 above)
Comment (reminia) [inline docs/usage/rest_api/notebook.md]: I'll change the name.

**Comment by reminia** [comment]
@zjffdu  @Leemoonsoo can you help review?

**Pull request closed without merging**